core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala

```diff
@@ -0,0 +1,202 @@
+    private val dfsLogFile: String = {
+      val rootDir = conf.get(DRIVER_LOG_DFS_DIR).get
+      if (!fileSystem.exists(new Path(rootDir))) {
+        throw new RuntimeException(s"${rootDir} does not exist." +
+          s" Please create this dir in order to persist driver logs")
+      }
+      FileUtils.getFile(rootDir, appId, DRIVER_LOG_FILE).getAbsolutePath()
+    }
+    private var inStream: InputStream = null
+    private var outputStream: FSDataOutputStream = null
+    try {
+      inStream = new BufferedInputStream(new FileInputStream(localLogFile))
+      outputStream = fileSystem.create(new Path(dfsLogFile), true)
+      fileSystem.setPermission(new Path(dfsLogFile), LOG_FILE_PERMISSIONS)
+    } catch {
+      case e: Exception =>
+        if (inStream != null) {
+          Utils.tryLogNonFatalError(inStream.close())
+        }
+        if (outputStream != null) {
+          Utils.tryLogNonFatalError(outputStream.close())
+        }
+        throw e
+    }
+    private val tmpBuffer = new Array[Byte](UPLOAD_CHUNK_SIZE)
+    private val threadpool = ThreadUtils.newDaemonSingleThreadScheduledExecutor("dfsSyncThread")
+    threadpool.scheduleWithFixedDelay(this, UPLOAD_INTERVAL_IN_SECS, UPLOAD_INTERVAL_IN_SECS,
+      TimeUnit.SECONDS)
+    logInfo(s"The local driver log file is being synced to spark dfs at: ${dfsLogFile}")
+
+    def run(): Unit = {
+      if (streamClosed) {
+        return
+      }
+      try {
+        var remaining = inStream.available()
+        while (remaining > 0) {
+          val read = inStream.read(tmpBuffer, 0, math.min(remaining, UPLOAD_CHUNK_SIZE))
+          outputStream.write(tmpBuffer, 0, read)
+          remaining -= read
+        }
+        outputStream.hflush()
+      } catch {
+        case e: Exception => logError("Failed writing driver logs to dfs", e)
+      }
+    }
+
+    private def close(): Unit = {
+      if (streamClosed) {
+        return
+      }
+      try {
+        // Write all remaining bytes
+        run()
+      } finally {
+        try {
+          streamClosed = true
+          inStream.close()
+          outputStream.close()
+        } catch {
+          case t: Throwable =>
+            logError("Error in closing driver log input/output stream", t)
+        }
+      }
+    }
+
+    def closeWriter(): Unit = {
+      try {
+        threadpool.execute(new Runnable() {
+          override def run(): Unit = DfsAsyncWriter.this.close()
+        })
+        threadpool.shutdown()
+        threadpool.awaitTermination(1, TimeUnit.MINUTES)
+      } catch {
+        case e: Exception =>
+          logError("Error in shutting down threadpool", e)
+      }
+    }
+  }
+
+}
+
+private[spark] object DriverLogger extends Logging {
+  val APPENDER_NAME = "_DriverLogAppender"
+
+  def apply(conf: SparkConf): Option[DriverLogger] = {
+    if (conf.get(DRIVER_LOG_PERSISTTODFS) && Utils.isClientMode(conf)) {
+      if (conf.get(DRIVER_LOG_DFS_DIR).isDefined) {
+        Try[DriverLogger] {
```

Comment:
This is actually simpler with a try..catch. There might have been a reason for this before but in it's current shape, a try..catch is simpler and actually easier to understand.